spring-batch-infrastructure/src/main/java/org/springframework/batch/item/json/JsonFileItemWriter.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.batch.item.json;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+
+import org.springframework.batch.item.WriteFailedException;
+import org.springframework.batch.item.WriterNotOpenException;
+import org.springframework.batch.item.support.AbstractFileItemWriter;
+import org.springframework.core.io.Resource;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Item writer that writes data in json format to an output file. The location
+ * of the output file is defined by a {@link Resource} and must represent a
+ * writable file. Items are transformed to json format using a
+ * {@link JsonObjectMarshaller}. Items will be enclosed in a json array as follows:
+ *
+ * <p>
+ * <code>
+ * [
+ *  {json object},
+ *  {json object},
+ *  {json object}
+ * ]
+ * </code>
+ * </p>
+ *
+ * The implementation is <b>not</b> thread-safe.
+ *
+ * @see GsonJsonObjectMarshaller
+ * @see JacksonJsonObjectMarshaller
+ * @param <T> type of object to write as json representation
+ * @author Mahmoud Ben Hassine
+ * @since 4.1
+ */
+public class JsonFileItemWriter<T> extends AbstractFileItemWriter<T> {
+
+	private static final char JSON_OBJECT_SEPARATOR = ',';
+	private static final char JSON_ARRAY_START = '[';
+	private static final char JSON_ARRAY_STOP = ']';
+
+	private JsonObjectMarshaller<T> jsonObjectMarshaller;
+
+	/**
+	 * Create a new {@link JsonFileItemWriter} instance.
+	 * @param resource to write json data to
+	 * @param jsonObjectMarshaller used to marshal object into json representation
+	 */
+	public JsonFileItemWriter(Resource resource, JsonObjectMarshaller<T> jsonObjectMarshaller) {
+		Assert.notNull(resource, "resource must not be null");
+		Assert.notNull(jsonObjectMarshaller, "json object marshaller must not be null");
+		setResource(resource);
+		setJsonObjectMarshaller(jsonObjectMarshaller);
+		setHeaderCallback(writer -> writer.write(JSON_ARRAY_START));
+		setFooterCallback(writer -> writer.write(this.lineSeparator + JSON_ARRAY_STOP + this.lineSeparator));
```

Comment:
by using a custom header/footer callback (see `org.springframework.batch.item.json.builder.JsonFileItemWriterBuilder#headerCallback` and `org.springframework.batch.item.json.builder.JsonFileItemWriterBuilder#footerCallback`)

Reply:
But if I specify a custom callback, who adds the opening `[` or closing `]` if I add a footer?  On the `StaxEventItemWriter`, we separate the writing of the header/footer from opening/closing the document.  I'd expect the same behavior here.  Specifically, if I understand how it is coded, if I configured a custom header callback then (as coded) I am responsible for adding the `[` which should not be the case.

Reply:
> who adds the opening [ or closing ]

No one adds them or should add them except the user because what if I don't want them at all? If I want to enclose my items with an JSON object (`{ items }`) instead of a JSON array (`[ items ]`), the writer should not add the `[` `]` on my behalf. Do you agree?

> On the StaxEventItemWriter, we separate the writing of the header/footer from opening/closing the document. I'd expect the same behavior here

The structure of an XML document is different, so IMO it should not the same behaviour. For XML we have:

```
- XML declaration
- begin root tag
- custom header
- items (one per line or pretty printed)
- custom footer
- end root tag
```

For JSON, it is more like a flat file (hence the inheritance) rather than XML:

```
- custom header
- items (one per line or pretty printed)
- custom footer
```

> if I configured a custom header callback then (as coded) I am responsible for adding the [ which should not be the case.

Yes, you are responsible for adding it because the writer cannot know (and should not) which enclosing character you want. In the trades example  (`org.springframework.batch.item.json.JsonFileItemWriterFunctionalTests#testJsonWritingWithEnclosingObject`), I specified a:

* custom header: `{"trades":[`
* custom footer: `]}`

and the result is:

```json
{"trades":[
 { "trade":1 },
 { "trade":2 },
 { "trade":3 }
]}
```

If I understand correctly , what you suggest is that I'm responsible for specifying a

* custom header:  `{"trades":` (without the opening `[`)
* custom footer: `}` (without the closing `]`)

and the writer would add them for me. This is an issue because if I don't want to enclose my items with `[` `]`, say like this (which is for instance [mongo's default format](https://docs.mongodb.com/manual/reference/program/mongoexport/#cmdoption-mongoexport-jsonarray)):

```json
 { "trade":1 },
 { "trade":2 },
 { "trade":3 }
```

the writer will add them for me which is not what I want.